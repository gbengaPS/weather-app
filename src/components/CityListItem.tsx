import React from 'react';
import { useHistory } from 'react-router-dom';
import { CityDetails, FavoritesMap } from '../types';
import { countriesMap } from '../constants';
import { getIconURL } from '../helpers';
import FavoriteButton from './FavoriteButton';



interface CityListItemProps {
  city: CityDetails
  onRemoveItem: Function
  favoritesMap: FavoritesMap
  onFavoriteClick: Function
}

const CityListItem = (props: CityListItemProps) => {
  const history = useHistory();
  const { city, onRemoveItem, favoritesMap, onFavoriteClick } = props;

  const handleCityClick = (city: string) => {
    history.push(`/${city}/details`);
  }

  return (
    <li key={city.name} >
      <div className="city-info" onClick={() => handleCityClick(city.name)}>
        <span>
          {city.name},
          <span className="bold-text">
            &nbsp;{countriesMap[city.sys.country]}
          </span>
        </span>
        <div className="aligned-flex">
          <span className="temperature">
            {Math.floor(city.main.temp)}&#8451;
              </span>
          <img src={getIconURL(city.weather[0].icon)} alt={city.weather[0].description} />
        </div>
      </div>
      <div className="">
        <button onClick={() => onRemoveItem(city)} className="app-button">
          <i className="fa fa-trash" aria-hidden="true"></i>
        </button>
        <FavoriteButton
          city={city}
          onFavoriteClick={onFavoriteClick}
          favoritesMap={favoritesMap}
        />
      </div>
    </li>
  );
};

export default CityListItem;