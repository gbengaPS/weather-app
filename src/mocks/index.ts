export const citiesMock = {
  base: "stations",
  clouds: {all: 0},
  cod: 200,
  coord: {lon: 116.4, lat: 39.91},
  dt: 1602226045,
  id: 1816670,
  main: {temp: 23.52, feels_like: 22.22, temp_min: 23.33, temp_max: 23.89, pressure: 1014, humidity: 47},
  name: "Beijing",
  sys: {type: 3, id: 2000022, country: 'CN', sunrise: 1602195502, sunset: 1602236679},
  timezone: 28800,
  visibility: 10000,
  weather: [{id: 800, main: "Clear", description: "clear sky", icon: "01d"}],
  wind: {speed: 2.55, deg: 209},
};

export const initialStateMock =  {
  favorites: { 'Beijing': { data: citiesMock } },
  weatherInfo: [],
  notes: { 'Beijing': 'hello world' },
  isLoading: false
}