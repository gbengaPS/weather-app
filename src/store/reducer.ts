const reducer = (state: any, action: any) => {
  return {}
}

export default reducer;