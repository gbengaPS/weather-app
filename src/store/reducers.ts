import { REMOVE_WEATHER_INFO, ADD_WEATHER_INFO, ADD_FAVORITE, REMOVE_FAVORITE } from '../constants';
import { FavoritesMap, WeatherInfo } from '../types';
import { saveFavorites } from '../helpers';

export const weatherInfoReducer = (state: WeatherInfo[], action: any) => {
  switch(action.type) {
    case REMOVE_WEATHER_INFO:
      const filteredInfo = state.filter((info: WeatherInfo) => {
        return info.data.name !== action.cityName;
      });
      localStorage.setItem('largestCitiesInfo', JSON.stringify(filteredInfo))
      return filteredInfo;

    case ADD_WEATHER_INFO:
      return [ ...state, ...action.weatherInfo ];
    default:
      return state;
  }
}

export const favoritesReducer = (state: FavoritesMap, action: any) => {
  switch(action.type) {
    case ADD_FAVORITE:
      const newState = { ...state, ...action.favoriteItem };
      saveFavorites(newState)
      return newState;

    case REMOVE_FAVORITE:
      const stateCopy = {...state};
      delete stateCopy[action.cityName];
      saveFavorites(stateCopy);
      return stateCopy;
    default:
      return state
  }
}