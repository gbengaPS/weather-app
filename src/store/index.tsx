import React, { createContext, useReducer } from 'react';
import { FavoritesMap, WeatherInfo } from '../types';
import { weatherInfoReducer, favoritesReducer } from './reducers';

interface InitialStateType {
	favorites: FavoritesMap
	weatherInfo: WeatherInfo[]
}

const initialState  = {
	favorites: {},
	weatherInfo: []
}

const AppContext = createContext<{
    state: InitialStateType;
    dispatch: React.Dispatch<any>;
  }>({
    state: initialState,
    dispatch: () => null
  });

const mainReducer = (state: InitialStateType, action: any) => ({
  weatherInfo: weatherInfoReducer(state.weatherInfo, action),
  favorites: favoritesReducer(state.favorites, action)
});

const AppProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(mainReducer, initialState);

  return (
      <AppContext.Provider value={{state, dispatch}}>
      {children}
      </AppContext.Provider>
  )
}
  
  export { AppContext, AppProvider };