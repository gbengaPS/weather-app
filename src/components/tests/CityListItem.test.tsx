import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import CityListItem from '../CityListItem';
import { AppContext } from '../../store';
import { citiesMock } from '../../mocks';

jest.mock('react-router-dom', () => ({
  useHistory: jest.fn(() => ({ push: jest.fn() }))
}));

describe('CityListItem', function() {
  afterEach(cleanup)

  afterAll(() => {
    jest.clearAllMocks();
  })

  const props = { city: citiesMock };

  it('should match snapshot', () => {
    const { container } = render(<CityListItem {...props}/>);
    expect(container).toMatchSnapshot();
  });


  it('should handle user click on list item', () => {
    const { getByText } = render(<CityListItem {...props}/>);

    fireEvent.click(getByText('Beijing,'));
  });

  it('should call dispatch twice when delete button is clicked', () => {
    const dispatch = jest.fn();
    const initialState = {
      favorites: {},
      weatherInfo: [],
      notes: {},
      isLoading: false
    }
    const { getByTestId } = render(
      <AppContext.Provider value={{dispatch, state: initialState }}>
        <CityListItem {...props}/>
      </AppContext.Provider>
    );

    fireEvent.click(getByTestId('delete-info'));
    expect(dispatch).toHaveBeenCalledTimes(2)
  });
});
