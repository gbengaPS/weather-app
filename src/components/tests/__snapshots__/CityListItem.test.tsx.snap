// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityListItem should match snapshot 1`] = `
<div>
  <li>
    <div
      class="city-info"
    >
      <span>
        Beijing
        ,
        <span
          class="bold-text"
        >
           
          China
        </span>
      </span>
      <div
        class="aligned-flex"
      >
        <span
          class="temperature"
        >
          23
          ℃
        </span>
        <img
          alt="clear sky"
          src="https://openweathermap.org/img/wn/01d@2x.png"
        />
      </div>
    </div>
    <div
      class=""
    >
      <button
        class="app-button"
        data-testid="delete-info"
      >
        <i
          aria-hidden="true"
          class="fa fa-trash"
        />
      </button>
      <button
        class="app-button"
        style="color: rgb(0, 0, 0);"
      >
        <i
          aria-hidden="true"
          class="fa fa-heart-o"
          data-testid="favorite-button"
        />
      </button>
    </div>
  </li>
</div>
`;
